### PR TITLE
add brave tool provider

### PR DESCRIPTION
**Pull request by dineshyv**
# What does this PR do?

Adds a new brave tool provider

## Test Plan

```
curl -X POST 'http://localhost:5000/alpha/toolgroups/register' \
-H 'Content-Type: application/json' \
-d '{
  "name": "search",
  "tool_group": {
    "type": "user_defined",
    "tools": [
      {
        "name": "brave_search",
        "description": "A web search tool",
        "parameters": [
          {
            "name": "query",
            "parameter_type": "string",
            "description": "The query to search"
          }
        ],
        "metadata": {},
        "tool_prompt_format": "json"
      }
    ]
  }
}'

curl -X POST 'http://localhost:5000/alpha/tool-runtime/invoke' \
-H 'Content-Type: application/json' \
-d '{
    "tool_id": "brave_search",
    "args": {
        "query": "who is meta ceo"
    }
}' | jq .content
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1973  100  1884  100    89  11288    533 --:--:-- --:--:-- --:--:-- 11885
"{'title': 'Mark Zuckerberg, Founder, Chairman and Chief Executive ...', 'url': 'https://about.meta.com/media-gallery/executives/mark-zuckerberg/', 'description': 'Not Logged In · Please log in to see this page', 'type': 'search_result'}\n{'title': 'Meta - Leadership & Governance', 'url': 'https://investor.fb.com/leadership-and-governance/', 'description': '<strong>Mark Zuckerberg</strong> is the founder, chairman and CEO of Meta, which he originally founded as Facebook in 2004. Mark is responsible for setting the overall direction and product strategy for the company. He leads the design of Meta&#x27;s services and development of its core technology and infrastructure.', 'type': 'search_result'}\n[{'type': 'video_result', 'url': 'https://www.facebook.com/etribune/videos/mark-zuckerberg-the-ceo-of-meta-has-officially-joined-the-exclusive-200-billion-/2372542949752515/', 'title': 'Mark Zuckerberg, the CEO of Meta, has officially joined the ...', 'description': \"Express Tribune, Karachi, Pakistan. 2,334,400 likes · 36,360 talking about this · 205 were here. The Express Tribune is Pakistan's #1 brand for breaking news in politics, sports, business, lifestyle\"}, {'type': 'video_result', 'url': 'https://www.youtube.com/watch?v=Y3oeQqtRvqk', 'title': \"Meta CEO: Mark Zuckerberg becomes World's Second Richest Person!\", 'description': 'Try VectorVest Risk-Free ➥➥➥ https://www.vectorvest.com/YTUse this link for a FREE Stock Analysis Report ➥➥➥ vectorvest.com/YTFSAVectorVest Merch Store ➥➥➥'}, {'type': 'video_result', 'url': 'https://m.facebook.com/abplive/videos/watch-meta-founder-and-ceo-mark-zuckerberg-recently-posted-a-heartwarming-video-/534841222497600/', 'title': '#WATCH | Meta founder and CEO Mark Zuckerberg recently ...', 'description': 'See posts, photos and more on Facebook'}]"

curl -X POST 'http://localhost:5000/alpha/tool-runtime/invoke' \
-H 'Content-Type: application/json' -H 'X-LlamaStack-ProviderData: {"api_key": "<KEY>"}' \
-d '{
    "tool_id": "brave_search",
    "args": {
        "query": "who is meta ceo"
    }
}'
```

### CODE REVIEW

File: llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        return
+
+    async def invoke_tool(self, tool_id: str, args: Dict[str, Any]) -> InvokeToolResult:
+        results = await self.execute(args["query"])
+        content_items = "\n".join([str(result) for result in results])
+        return InvokeToolResult(
+            content=content_items,
+        )
+
+    async def execute(self, query: str) -> List[dict]:
```
Comment (ashwinb) [inline llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py:38]: you could inline this I guess?

File: llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+import requests
+
+from llama_stack.apis.tools import InvokeToolResult, Tool, ToolRuntime
+from llama_stack.providers.datatypes import ToolsProtocolPrivate
+
+from .config import BraveSearchToolConfig
+
+
+class BraveSearchToolRuntimeImpl(ToolsProtocolPrivate, ToolRuntime):
```
Comment (ashwinb) [inline llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py:17]: you can now add `NeedsProviderData` mixin here and request the API_KEY through the provider_data mechanism we have. Look at the Together inference impl.

**Review by ashwinb — APPROVED** [review]

File: llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py (hunk 1 above)
Comment (dineshyv) [reply llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py]: done.

File: llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py (hunk 2 above)
Comment (dineshyv) [reply llama_stack/providers/inline/tool_runtime/brave_search/brave_search.py]: done.

**Pull request merged**